packages/@aws-cdk/aws-redshift-alpha/lib/user.ts

```diff
@@ -153,6 +153,7 @@ export class User extends UserBase {
     const secret = new DatabaseSecret(this, 'Secret', {
       username,
       encryptionKey: props.encryptionKey,
+      masterSecret: props.adminUser,
```

Comment:
Not sure about this.
Shouldn't we simply pass a `DatabaseSecret` with `masterSecret` set to the `addRotationMultiUser` method?
It seems like we might be assigning unnecessary permissions to the user here.
Also, I think we should validate that the secret for `addRotationMultiUser` has the required `masterSecret` value set [here](https://github.com/aws/aws-cdk/blob/95d187e4aabf69b7210ffdf53e53eda077b46d6e/packages/%40aws-cdk/aws-redshift-alpha/lib/cluster.ts#L653).

Reply:
Absolutely! I was having the same conversation with a teammate this morning about that. The secret itself shouldn't need to encode the masterSecret reference especially since there is nothing secret about that reference (nor anything bad that could happen if an incorrect or different arn was set - the rotation lambda would just fetch wrong cluster admin credentials and fail to make the update!).

I think we may be bound by the implementation of [addRotationMultiUser](https://github.com/aws/aws-cdk/blob/v2.122.0/packages/%40aws-cdk/aws-redshift-alpha/lib/cluster.ts#L654C31-L662) and the [redshift multi-user rotation lambda](https://github.com/aws/aws-cdk/blob/v2.122.0/packages/%40aws-cdk/aws-redshift-alpha/lib/cluster.ts#L534) secrets manager provides.

Re unassigning unecessary permissions, I see your point about sending `masterSecret` into DatabaseSecret but at impl time its just encoding the master secret ARN w/o additional permissions. Its the rotation lambda that has permission to the masterSecret and secret to perform the rotation.

On the validation, I don't know if we can perform that validation at build time since `options.secret` is a reference to ISecret and not DatbaseSecret :/ - thoughts?

Reply:
Makes sense and I agree with your analysis of the current implementation.
Thanks for the detailed follow-up 👍
On validation, we could expose the `masterSecret` property publicly in `DatabaseSecret` and perform a type-check in `addRotationMultiUser` for the `secret`.
I think it might be worth adding to provide clearer feedback to users invoking the method without providing the necessary parameters.